### PR TITLE
initramfs-init: add support for ssh installer

### PR DESCRIPTION
**Pull request by tmh1999**
This commit allows running the Alpine installer through SSH either
using a SSH key or a password. Allowing root access through SSH must
be revoked before finishing running setup-disk as this config would
leak to new system.

On s390x, the boot media consist of a kernel, an initramfs, and a
parm file. We don't necessarily need a modloop image and fetch
initial packages from remote repo instead. This behavior is the same
on other distros.

### CODE REVIEW

**Comment by tmh1999** [comment]
This is the fix for 2nd commit of https://github.com/alpinelinux/mkinitfs/pull/26

File: initramfs-init.in [hunk 1]
```diff
@@ -269,7 +269,7 @@ set -- $(cat /proc/cmdline)
 myopts="alpine_dev autodetect autoraid chart cryptroot cryptdm cryptheader cryptoffset
 	cryptdiscards debug_init dma init_args keep_apk_new modules ovl_dev pkgs quiet
 	root_size root usbdelay ip alpine_repo apkovl alpine_start splash blacklist
-	overlaytmpfs rootfstype rootflags nbd resume s390x_net dasd"
+	overlaytmpfs rootfstype rootflags nbd resume s390x_net dasd nameserver ssh_install ssh_key ssh_pass"
```
Comment (ncopa) [inline initramfs-init.in]: I think we don't need `nameserver` which can/should be passed via `ip`, similar to how it is done for [nfsroot](https://www.kernel.org/doc/Documentation/filesystems/nfs/nfsroot.txt). I also don't think we need the `ssh_install`. We can enable ssh server if either `ssh_key` or `ssh_pass` is set.

File: initramfs-init.in [hunk 2]
```diff
@@ -364,6 +364,11 @@ if [ "${s390x_net%%,*}" = "qeth_l2" ]; then
 	echo 1 > /sys/bus/ccwgroup/drivers/qeth/"${_channel%%,*}"/online
 fi
 
+if [ -n "$ssh_install" ]; then
```
Comment (ncopa) [inline initramfs-init.in]: How about we drop the `ssh_install` and instead do:
````sh
if [ -n "$ssh_key" ] || [ -n "$ssh_pass" ]; then
    ...
````

File: initramfs-init.in [hunk 3]
```diff
@@ -364,6 +364,11 @@ if [ "${s390x_net%%,*}" = "qeth_l2" ]; then
...
+	echo "nameserver $nameserver" > /etc/resolv.conf
```
Comment (ncopa) [inline initramfs-init.in]: This is buggy. You probably mean `$KOPT_nameserver` here. But we may need to set the name server even if `ssh_install` (or `ssh_key`/`ssh_pass`) is unset, so it cannot be within the conditional. A better place may be to add it to the `ip=....` variable.

File: initramfs-init.in [hunk 4]
```diff
@@ -364,6 +364,11 @@ if [ "${s390x_net%%,*}" = "qeth_l2" ]; then
...
+	ALPINE_REPO=${KOPT_alpine_repo:-http://dl-cdn.alpinelinux.org/alpine/latest-stable/main}
```
Comment (ncopa) [inline initramfs-init.in]: This is will not work if someone has `alpine_repo=auto` at the command line. We should probably do something like:
````sh
if [ -n "$ssh_key" ] || [ -n "$ssh_pass" ]; then
    : ${ALPINE_REPO:=http://dl-cdn.alpinelinux.org/alpine/latest-stable/main}
fi
````
`ALPINE_REPO` will be empty if user set `alpine_repo=auto` boot option.

File: initramfs-init.in [hunk 5]
```diff
@@ -591,6 +596,7 @@ fi
 
 # hack so we get openrc
 pkgs="$pkgs alpine-base"
+[ -n "$ssh_install" ] && pkgs="$pkgs openssh"
```
Comment (ncopa) [inline initramfs-init.in]: How about:
````sh
if [ -n "$ssh_key" ] || [ -n "$ssh_pass" ]; then
    pkgs="$pkgs openssh"
fi
````

File: initramfs-init.in [hunk 6]
```diff
@@ -646,6 +652,22 @@ if [ -n "$ovlfiles" ]; then
 else
 	apk add --root $sysroot $repo_opt $apkflags $pkgs
 fi
+
+if [ -n "$ssh_install" ]; then
+	rc_add sshd default
+	echo "PermitRootLogin yes" >> $sysroot/etc/ssh/sshd_config
+	if [ -n "$ssh_key" ]; then
+		mkdir -p $sysroot/root/.ssh
+		wget "$ssh_key" -O $sysroot/root/.ssh/authorized_keys
```
Comment (ncopa) [inline initramfs-init.in]: How about:
````sh
￼if [ -n "$ssh_key" ]; then
￼    mkdir -p $sysroot/root/.ssh
    case "$ssh_key" in
        http://*|https://*) wget "$ssh_key" -O $sysroot/root/.ssh/authorized_keys;;
        *) echo "$ssh_key" > "$sysroot"/root/.ssh/authorized_keys;;
    esac
fi
````

That way you can also paste in your ed25519 key in the boot option.

File: initramfs-init.in [hunk 7]
```diff
@@ -646,6 +652,22 @@ if [ -n "$ovlfiles" ]; then
...
+	elif [ -n "$ssh_pass" ]; then
```
Comment (ncopa) [inline initramfs-init.in]: we shouldnt use `elif` here. I think it should be possible to set both `ssh_key` and `ssh_pass` if that is wanted.

File: initramfs-init.in [hunk 8]
```diff
@@ -646,6 +652,22 @@ if [ -n "$ovlfiles" ]; then
...
+		echo "root:$ssh_pass" > $sysroot/.passwd
+		chroot $sysroot sh -c 'chpasswd < /.passwd'
+		rm $sysroot/.passwd
```
Comment (ncopa) [inline initramfs-init.in]: Is it not possible to pipe directly to `chpasswd`?
````sh
    echo "root:$ssh_pass" | chroot $sysroot /usr/sbin/chpasswd
````

File: initramfs-init.in [hunk 9]
```diff
@@ -646,6 +652,22 @@ if [ -n "$ovlfiles" ]; then
...
+	fi
+	# to make modloop happy
+	mkdir -p $sysroot/.modloop/modules/$(uname -r)
```
Comment (ncopa) [inline initramfs-init.in]: Is this something that is only affecting openssh? I would believe it affects everything that uses `ALPINE_REPOS` and not only openssh.

**Review by ncopa — CHANGES_REQUESTED** [review]

**Comment by ncopa** [comment]
The `nameserver` can be replaced with this:
````diff
diff --git a/initramfs-init.in b/initramfs-init.in
index 933c8c1..2f0d6e6 100755
--- a/initramfs-init.in
+++ b/initramfs-init.in
@@ -183,6 +183,8 @@ configure_ip() {
        local netmask="$4"
        local device="$6"
        local autoconf="$7"
+       local dns1="$8"
+       local dns2="$9"
        case "$client_ip" in
                off|none|'') return;;
                dhcp) autoconf="dhcp";;
@@ -209,6 +211,8 @@ configure_ip() {
                ebegin "Setting IP ($device)..."
                ip_set "$device" "$client_ip" "$netmask" "$gw_ip"
                eend $?
+               [ -n "$dns1" ] && echo "nameserver $dns1" >> /etc/resolv.conf
+               [ -n "$dns2" ] && echo "nameserver $dns2" >> /etc/resolv.conf
        fi
        MAC_ADDRESS=$(cat /sys/class/net/$device/address)
 }
````

File: initramfs-init.in (hunk 1 above)
Comment (tmh1999) [reply initramfs-init.in]: Fixed.

File: initramfs-init.in (hunk 2 above)
Comment (tmh1999) [reply initramfs-init.in]: Fixed.

File: initramfs-init.in (hunk 4 above)
Comment (tmh1999) [reply initramfs-init.in]: This is helpful. Fixed.

File: initramfs-init.in (hunk 5 above)
Comment (tmh1999) [reply initramfs-init.in]: Fixed.

File: initramfs-init.in (hunk 6 above)
Comment (tmh1999) [reply initramfs-init.in]: That's nice. Fixed. Thanks.

File: initramfs-init.in (hunk 7 above)
Comment (tmh1999) [reply initramfs-init.in]: Fixed.

File: initramfs-init.in (hunk 8 above)
Comment (tmh1999) [reply initramfs-init.in]: Yeah I struggled a while for this... Did not know we can pipe to chroot. Fixed. Thanks.

File: initramfs-init.in (hunk 9 above)
Comment (tmh1999) [reply initramfs-init.in]: Well it's more like s390x stuff since s390x installer has only initrd without modloop (other arch installing through SSH may have modloop though). I remove this for now and revisit it later with better approach maybe.

Comment (tmh1999) [reply initramfs-init.in]: It's not critical though, just to make modloop init script not complainig during booting the installer.

File: initramfs-init.in [hunk 10]
```diff
@@ -629,7 +640,7 @@ else
 	apkflags="$apkflags --update-cache"
 fi
 
-if [ "$KOPT_quiet" = yes ]; then
+if [ "$KOPT_quiet" = yes ] || [ -n "$KOPT_ssh_key" ] || [ -n "$KOPT_ssh_pass" ]; then
```
Comment (ncopa) [inline initramfs-init.in]: Why is this needed?

Comment (tmh1999) [reply initramfs-init.in]: Because when we pull initial packages from the network, the output is a lot. Not a functional requirement though, I could remove it if desired.

Comment (tmh1999) [reply initramfs-init.in]: Alright this looks more like a s390x thing rather than SSH thing. I will come up with a way to make this work. Remove this now.

**Review by ncopa — APPROVED** [review]

File: initramfs-init.in [hunk 11]
```diff
@@ -209,6 +211,8 @@ configure_ip() {
 		ebegin "Setting IP ($device)..."
 		ip_set "$device" "$client_ip" "$netmask" "$gw_ip"
 		eend $?
+		[ -n "$dns1" ] && echo "nameserver $dns1" >> /etc/resolv.conf
+		[ -n "$dns2" ] && echo "nameserver $dns2" >> /etc/resolv.conf
```
Comment (clandmeter) [inline initramfs-init.in:215]: Nice, i thought we already had this. If you want you can also move this to a separate commit and ill push it.

File: initramfs-init.in [hunk 12]
```diff
@@ -345,25 +349,29 @@ mount -t devpts -o gid=5,mode=0620,noexec,nosuid devpts /dev/pts
...
 	echo "$_channel" > /sys/bus/ccwgroup/drivers/qeth/group
 	echo 1 > /sys/bus/ccwgroup/drivers/qeth/"${_channel%%,*}"/layer2
 	echo 1 > /sys/bus/ccwgroup/drivers/qeth/"${_channel%%,*}"/online
 fi
 
+if [ -n "$KOPT_ssh_key" ] || [ -n "$KOPT_ssh_pass" ]; then
+	: ${ALPINE_REPO:=http://dl-cdn.alpinelinux.org/alpine/latest-stable/main}
+fi
+
 # load available drivers to get access to modloop media
```
Comment (clandmeter) [inline initramfs-init.in]: I don't like alpine_repo here. IF we set a global default we should do it more general and don't start to repeat it for every network option. I also wonder if we really need to use dl-cdn as the default repo. I think having a more generic name would be more suitable ie main.a.o or dl-default (or any good alternative).

File: initramfs-init.in [hunk 13]
```diff
@@ -646,6 +657,23 @@ if [ -n "$ovlfiles" ]; then
 else
 	apk add --root $sysroot $repo_opt $apkflags $pkgs
 fi
+
+if [ -n "$KOPT_ssh_key" ]; then
+	rc_add sshd default
+	echo "PermitRootLogin prohibit-password" >> $sysroot/etc/ssh/sshd_config
+	mkdir -p $sysroot/root/.ssh
+	case "$KOPT_ssh_key" in
+		http://*|https://*) wget "$KOPT_ssh_key" -O $sysroot/root/.ssh/authorized_keys;;
```
Comment (clandmeter) [inline initramfs-init.in]: As reported on [bugs](https://bugs.alpinelinux.org/issues/8917) we should depend on GNU wget instead of using busybox wget. GNU wget should be the default for whenever we want to fetch from a secure server (ie modloop and others).
And what happens here if i use ftp(s)?

**Review by clandmeter — CHANGES_REQUESTED** [review]
This PR name is confusing as you would assume it would init/start the installer which it doesn't. I thought about this now and i think it does make more sense to move this out of initramfs. I also like the idea that ncopa mentioned about first_boot. We could move this to a dedicated initd script called first_boot which we can use for other "first boot" related tasks. It would always run on first boot and do tasks depended on cmdline. It would be added to default run level and be removed whenever you do a sys install (like modloop) or when an ovl file is found. This will keep the init file smaller/simpler and we don't need to have wget in our initramfs.

Did you think about a way to remove the sshd_config allow password login?

File: initramfs-init.in (hunk 13 above)
Comment (tmh1999) [reply initramfs-init.in]: Updated everything in new commit. Thanks.
